contributors/design-proposals/sig-cli/kubectl_apply_getsetdiff_last_applied_config.md

```diff
@@ -0,0 +1,193 @@
+# Kubectl apply subcommands for last-config
+
+## Abstract
+
+`kubectl apply` uses the `last-applied-config` annotation to compute
+the removal of fields from local object configuration files and then
+send patches to delete those fields from the live object.  Reading or
+updating the `last-applied-config` is complex as it requires parsing
+out and writing to the annotation.  Here we propose a set of porcelain
+commands for users to better understand what is going on in the system
+and make updates.
+
+## Motivation
+
+What is going on behind the scenes with `kubectl apply` is opaque.  Users
+have to interact directly with annotations on the object to view
+and make changes.  In order to stop having `apply` manage a field on
+an object, it must be manually removed from the annotation and then be removed
+from the local object configuration.  Users should be able to simply edit
+the local object configuration and set it as the last-applied-config
+to be used for the next diff base.  Storing the last-applied-config
+in an annotation adds black magic to `kubectl apply`, and it would
+help users learn and understand if the value was exposed in a discoverable
+manner.
+
+## Use Cases
+
+1. As a user, I want to be able to diff the last-applied-configuration
+   against the current local configuration to see which changes the command is seeing
+2. As a user, I want to remove fields from being managed by the local
+   object configuration by removing them from the local object configuration
+   and setting the last-applied-configuration to match.
+3. As a user, I want to be able to view the last-applied-configuration
+   on the live object that will be used to calculate the diff patch
+   to update the live object from the configuration file.
+
+## Naming and Format possibilities
+
+### Naming
+
+1. last-applied
+
+Rejected alternatives:
+
+2. ~~last-config~~
+3. ~~last-applied-config~~
+4. ~~last-configuration~~
+5. ~~last-applied-configuration~~
+6. ~~last~~
+
+### Formats
+
+1. Apply subcommands
+  - `kubectl apply set/view/diff last-applied
+  - a little bit odd to have 2 verbs in a row
```

Comment:
Yes, something like `kubectl apply view-last-applied` was what I was thinking. 

In https://github.com/kubernetes/kubernetes/pull/41146 I originally suggested adding this feature in `kubectl get` or `kubectl describe` (maybe with a flag), since this command allows users to view resources instead of mutating them. However, after reading this proposal, I agree that discoverability is probably more important for this specific feature. Maybe just documenting "It's also a bit odd to view something with `kubectl apply`, which is supposed to update (not view) resources" in the proposal is enough (showing that we're aware of this but choose discoverability over this.)